web_poet/serialization/functions.py

```diff
@@ -0,0 +1,108 @@
+import json
+from typing import Type
+
+from .. import HttpResponse, HttpResponseBody, HttpResponseHeaders, ResponseUrl
+from ..page_inputs.url import _Url
+from .api import (
+    SerializedLeafData,
+    deserialize_leaf,
+    register_serialization,
+    serialize_leaf,
+)
+
+
+def _serialize_dict(o: dict) -> SerializedLeafData:
+    return {"json": json.dumps(o).encode()}
+
+
+def _deserialize_dict(t: Type[dict], data: SerializedLeafData) -> dict:
+    return t(json.loads(data["json"]))
+
+
+register_serialization(_serialize_dict, _deserialize_dict)
+
+
+def _serialize_HttpResponse(o: HttpResponse) -> SerializedLeafData:
+    other_data = {
+        "url": str(o.url),
+        "status": o.status,
+        "headers": list(o.headers.items()),
+        "_encoding": o._encoding,
+    }
+    body_serialized = serialize_leaf(o.body)
+    other_serialized = serialize_leaf(other_data)
+    result = {}
+    result.update({f"body.{k}": v for k, v in body_serialized.items()})
+    result.update({f"other.{k}": v for k, v in other_serialized.items()})
+    return result
+
+
+def _deserialize_HttpResponse(
+    t: Type[HttpResponse], data: SerializedLeafData
+) -> HttpResponse:
+    body_serialized = {}
+    other_serialized = {}
+    for k, v in data.items():
+        if k.startswith("body."):
+            body_serialized[k[len("body.") :]] = v
+        elif k.startswith("other."):
+            other_serialized[k[len("other.") :]] = v
+
+    body = deserialize_leaf(HttpResponseBody, body_serialized)
+    other_data = deserialize_leaf(dict, other_serialized)
+    return t(
+        body=body,
+        url=ResponseUrl(other_data["url"]),
+        status=other_data["status"],
+        headers=HttpResponseHeaders(other_data["headers"]),
+        encoding=other_data["_encoding"],
+    )
+
+
+register_serialization(_serialize_HttpResponse, _deserialize_HttpResponse)
+
+
+def _serialize_HttpResponseBody(o: HttpResponseBody) -> SerializedLeafData:
+    return {"html": bytes(o)}
+
+
+def _deserialize_HttpResponseBody(
+    t: Type[HttpResponseBody], data: SerializedLeafData
+) -> HttpResponseBody:
+    return t(data["html"])
+
+
+register_serialization(_serialize_HttpResponseBody, _deserialize_HttpResponseBody)
+
+
+def _serialize_bytes(o: bytes) -> SerializedLeafData:
+    return {"bin": bytes(o)}
+
+
+def _deserialize_bytes(t: Type[bytes], data: SerializedLeafData) -> bytes:
+    return t(data["bin"])
+
+
+register_serialization(_serialize_bytes, _deserialize_bytes)
+
+
+def _serialize_str(o: str) -> SerializedLeafData:
+    return {"txt": o.encode()}
+
+
+def _deserialize_str(t: Type[str], data: SerializedLeafData) -> str:
+    return t(data["txt"].decode())
+
+
+register_serialization(_serialize_str, _deserialize_str)
```

Comment:
Could you please group "primitive" functions, vs functions for concrete dependencies defined by web-poet?

Also, I think it might be good to add some comment about it, on why we need ser/de for str, dict, bytes. My understanding is that we don't really need it (as these types are very unlikely to be leaf dependencies), but we provide it so that all subclasses of these primitive classes don't require registering of custom functions. Is it correct?

There could be an argument for not doing this, and just creating functions for each dependency. But I don't have a strong opinon.

Reply:
Also, just to double-check: is this all working ok if there are 2 subclasses of the same primitive type (e.g. two str subclasses), and both are used? I assume that's working fine.

Reply:
Functions for `str` and `bytes` are indeed provided for the sake of completion and are not used by the existing code. 

The intended use cases for these (and other similar) primitive type functions are both to cover subclasses (though `HttpResponseBody` is a subclass of `bytes` and we provide functions for it, but just to have a better file extension) and to be able to write `serialize_leaf()` in functions like `_(de)serialize_HttpResponse()`. Though it may be easier to just inline all that code in the second use case.

> is this all working ok if there are 2 subclasses of the same primitive type (e.g. two str subclasses), and both are used? I assume that's working fine.

What use case do you have in mind? Registering functions for both subclasses should work correctly for both `serialize_leaf()` and `deserialize_leaf()`, and just using the functions provided for `str` should also work correctly. If you mean using both in a dependency list, then it should never matter that they are subclasses of `str`.

Reply:
> What use case do you have in mind?

Let's say we register `str`, and expect it to cover two subclasses: BrowserHtml(str) and MyBrowserHtml(str), without registering anything for these subclasses explicitly. My question was if this works correctly, but it seems it does, as the keys in dictionary are subclass names, and there is no clash.

Overall, I'm not 100% sure, but slightly prefer removing str, bytes, dict serializers, and supporting classes which we provide directly instead. I'm not sure what's the use case for making some arbitrary str or dict subclasses serializable, in context of page objects.

Reply:
I've removed those and inlined them in the HttpResponse implementation. I think it looks fine.

Note that now we handle HttpResponseBody manually instead of using its specific methods even though they still exists.